swirl/src/registry.rs

```diff
@@ -1,46 +1,96 @@
 #![allow(clippy::new_without_default)] // https://github.com/rust-lang/rust-clippy/issues/3632
 
-use serde_json;
+use std::any::{Any, TypeId};
 use std::collections::HashMap;
-use std::panic::RefUnwindSafe;
+use std::marker::PhantomData;
 
 use crate::errors::PerformError;
 use crate::Job;
 
-#[doc(hidden)]
-pub type PerformFn<Env> =
-    Box<dyn Fn(serde_json::Value, &Env) -> Result<(), PerformError> + RefUnwindSafe + Send + Sync>;
-
 #[derive(Default)]
 #[allow(missing_debug_implementations)] // Can't derive debug
-/// A registry of background jobs, used to map job types to concrege perform
+/// A registry of background jobs, used to map job types to concrete perform
 /// functions at runtime.
 pub struct Registry<Env> {
-    job_types: HashMap<&'static str, PerformFn<Env>>,
+    jobs: HashMap<&'static str, JobVTable>,
+    _marker: PhantomData<Env>,
 }
 
-impl<Env> Registry<Env> {
-    /// Create a new, empty registry
-    pub fn new() -> Self {
-        Registry {
-            job_types: Default::default(),
+impl<Env: 'static> Registry<Env> {
+    /// Loads the registry from all invocations of [`register_job!`] for this
+    /// environment type
+    pub fn load() -> Self {
+        let jobs = inventory::iter::<JobVTable>
+            .into_iter()
+            .filter(|vtable| vtable.env_type == TypeId::of::<Env>())
+            .map(|&vtable| (vtable.job_type, vtable))
+            .collect();
+
+        Self {
+            jobs: jobs,
+            _marker: PhantomData,
         }
     }
 
     /// Get the perform function for a given job type
-    pub fn get(&self, job_type: &str) -> Option<&PerformFn<Env>> {
-        self.job_types.get(job_type)
+    pub fn get(&self, job_type: &str) -> Option<PerformJob<Env>> {
+        self.jobs.get(job_type).map(|&vtable| PerformJob {
+            vtable,
+            _marker: PhantomData,
+        })
     }
+}
+
+/// Register a job to be run by swirl. This must be called for any
+/// implementors of [`swirl::Job`]
+#[macro_export]
+macro_rules! register_job {
+    ($job_ty: ty) => {
+        $crate::inventory::submit! {
+            #![crate = swirl]
+            swirl::JobVTable::from_job::<$job_ty>()
```

Comment:
There's a regression in 1.32 which prevents using `$crate` here, I'm trying to minimize this to a smaller example to report to rustc. Using the actual crate name here is fine, but requires that the crate isn't renamed in Cargo.toml (which is already going to be a requirement since we're planning on doing proc_macro codegen), and that the downstream crate is using the 2018 edition.